activesupport/lib/active_support/rescuable.rb

```diff
@@ -65,27 +63,67 @@ def rescue_from(*klasses, &block)
           end
 
           # Put the new handler at the end because the list is read in reverse.
-          self.rescue_handlers += [[key, options[:with]]]
+          self.rescue_handlers += [[key, with]]
         end
       end
-    end
 
-    # Tries to rescue the exception by looking up and calling a registered handler.
-    def rescue_with_handler(exception)
-      if handler = handler_for_rescue(exception)
-        handler.arity != 0 ? handler.call(exception) : handler.call
-        true # don't rely on the return value of the handler
+      # Matches an exception to a handler based on the exception class.
+      #
+      # If no handler matches the exception, check for a handler that matches
+      # the exception.cause, if present. (An exception's *cause* is the
+      # original exception that was being rescued when this exception was
+      # raised. The original exception *caused* this one since we raised this
+      # while rescuing the other.)
+      #
+      # If no handler matches the exception or its cause, this does nothing
+      # and returns `nil` so you can deal with unhandled exceptions. Be sure
+      # to re-raise unhandled exceptions if this is what you expect:
+      #
+      #     begin
+      #       …
+      #     rescue => exception
+      #       rescue_with_handler(exception) || raise
+      #     end
+      def rescue_with_handler(exception, object: self)
+        if handler = handler_for_rescue(exception, object: object)
+          handler.arity != 0 ? handler.call(exception) : handler.call
+          true # Always returns true if the exception was handled.
+        end
       end
-    end
 
-    def handler_for_rescue(exception)
-      # We go from right to left because pairs are pushed onto rescue_handlers
-      # as rescue_from declarations are found.
-      _, rescuer = self.class.rescue_handlers.reverse.detect do |klass_name, handler|
+      def handler_for_rescue(exception, object: self) #:nodoc:
+        case rescuer = find_rescue_handler(exception)
+        when Symbol
+          object.method(rescuer)
+        when Proc
+          if rescuer.arity == 0
+            Proc.new { object.instance_exec(&rescuer) }
+          else
+            Proc.new { |_exception| object.instance_exec(_exception, &rescuer) }
+          end
+        end
+      end
+
+      private
+        def find_rescue_handler(exception)
+          if exception
+            # Handlers are in order of declaration but the most recently declared
+            # is the highest priority match, so we search for matching handlers
+            # in reverse.
+            _, handler = rescue_handlers.reverse_each.detect do |class_name, _|
+              if klass = constantize_rescue_handler_class_name(class_name)
+                klass === exception
+              end
+            end
+
+            handler || find_rescue_handler(exception.cause)
+          end
+        end
```

Comment:
Find a handler that matches the exception, recursing to exception.cause if none match.